include/sparrow/c_interface.hpp

```diff
@@ -0,0 +1,437 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdlib>
+#include <iterator>
+#include <memory>
+#include <optional>
+#include <span>
+#include <utility>
+
+#include "sparrow/allocator.hpp"
+#include "sparrow/contracts.hpp"
+
+
+extern "C"
+{
+    struct ArrowSchema
+    {
+        // Array type description
+        const char* format = nullptr;
+        const char* name = nullptr;
+        const char* metadata = nullptr;
+        int64_t flags = 0;
+        int64_t n_children = 0;
+        struct ArrowSchema** children = nullptr;
+        struct ArrowSchema* dictionary = nullptr;
+
+        // Release callback
+        void (*release)(struct ArrowSchema*) = nullptr;
+        // Opaque producer-specific data
+        void* private_data = nullptr;
+    };
+
+    struct ArrowArray
+    {
+        // Array data description
+        int64_t length = 0;
+        int64_t null_count = 0;
+        int64_t offset = 0;
+        int64_t n_buffers = 0;
+        int64_t n_children = 0;
+        const void** buffers = nullptr;
+        struct ArrowArray** children = nullptr;
+        struct ArrowArray* dictionary = nullptr;
+
+        // Release callback
+        void (*release)(struct ArrowArray*) = nullptr;
+        // Opaque producer-specific data
+        void* private_data = nullptr;
+    };
+
+}  // extern "C"
+
+namespace sparrow
+{
+    template <class T>
+    std::vector<T*> to_raw_ptr_vec(const std::vector<std::unique_ptr<T>>& vec)
+    {
+        std::vector<T*> raw_ptr_vec;
+        std::ranges::transform(
+            vec,
+            std::back_inserter(raw_ptr_vec),
+            [](const auto& child)
+            {
+                return child.get();
+            }
+        );
+        return raw_ptr_vec;
+    }
+
+    template <class T>
+    std::vector<T*> to_raw_ptr_vec(std::vector<std::vector<T>>& vectors)
```

Comment:
I think you need an additional template parameter for the allocator of the internal `vector` here (given the way you use it in the `ArrowArrayPrivateData` structure).